### DISCUSSION TITLE
How to make Relations with Custom Models

### DISCUSSION THREAD

**Discussion by Jangerlp** [pocketbase/pocketbase#1819] [Q&A] [answered]
I have two tables (teams and players) where a team has multiple members (ülayers)
How do I set up the relation correctly?

Player:
```go
type Player struct {
	models.BaseModel

	Name            string `db:"name" json:"name"`
	Mmr             int    `db:"mmr" json:"mmr"`
	ProfileId       int    `db:"profileId" json:"profileId"`
}
```

Team:
```go
type Team struct {
	models.BaseModel

	Members    []string `db:"members" json:"members"` //<---- Should relate to User
	Mmr        int      `db:"mmr" json:"mmr"`
	OwnTeam    bool     `db:"ownTeam" json:"ownTeam"`
}
```

I'm saving them to the databse using the save function
```go
err := app.Dao().Save(team)
```

And this is how my teams collection is initialized:
```go
var teamsCollection = &models.Collection{
	Name:       "teams",
	Type:       models.CollectionTypeBase,
	ListRule:   nil,
	ViewRule:   nil,
	CreateRule: nil,
	UpdateRule: nil,
	DeleteRule: nil,
	Schema: schema.NewSchema(
		&schema.SchemaField{
			Name:     "members",
			Type:     schema.FieldTypeRelation,
			Required: true,
			Unique:   false,
			Options: &schema.RelationOptions{
				CollectionId:  playersCollection.GetId(),
				CascadeDelete: true,
			},
		},
		//...
	),
}
```

**Comment by ganigeorgiev** [comment]
Relations are referenced and stored only by their id, aka. there is no builtin mechanism that will magically populate `Player` structs as part of the `Team.Member` field, if that's what you are looking for.

The question is too generic and lack context and I'm not sure what to suggest, but if you want to fetch the related `Player` of a `Team`, the simplest way would be to create a generic helper like:

```go
func MemberQuery(dao *daos.Dao) *dbx.SelectQuery {
    return dao.ModelQuery(&Member{})
}


func FindMembers(dao *daos.Dao, memberIds ...any) ([]*Member, error) {
    members := make([]*Member, 0, len(memberIds))

    err := MemberQuery(dao).
    	AndWhere(dbx.In("id", memberIds...)).
        All(&members)

    if err != nil {
        return nil, err
    }

    return members, nil
}

members, err := FindMembers(app.Dao(), team.Members...)
```

If you are asking how to save the `[]string` in the database, then it must be as a serialized json array. You can use the [`types.JsonArray` struct](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.12.2/tools/types#JsonArray) instead of `[]string` since it already implements the `driver.Valuer` and `sql.Scanner` interfaces.

```go
type Team struct {
	models.BaseModel

	Members types.JsonArray `db:"members" json:"members"`
	Mmr     int             `db:"mmr" json:"mmr"`
	OwnTeam bool            `db:"ownTeam" json:"ownTeam"`
}

team := new(Team)
team.Members = types.JsonArray{"id1", "id2", "etc."}

if err := app.Dao().Save(team); err != nil {
	return err
}
```

---

But once again, I'm not sure what you are trying to do and the above may not be suitable for your use case.

Why do you want to use custom models in the first place? Is it only for the type safety? Is the above part of a custom endpoint?

Note that the recommended way to interact with your collections data is usually using the [`Record` model and its helpers](https://pocketbase.io/docs/record-methods/) since the app event hooks works with them out of the box. The support for custom models was added primarily to allow interacting with other non-collections data, aggregations, or to create a composite of multiple related structures from different places (_you can query multiple collections at once and automatically populate the fields by prefixing the selected column statement with the related table name; you can find an example in the dbx package or in https://github.com/pocketbase/pocketbase/discussions/1637#discussioncomment-4720884_).

**Reply by Jangerlp** [reply]
Thanks for your answer

A little bit of context, I have a game which produces an xml file with statistics after every match, I take this xml file and read out all the teams, players and events and stick them together into the database where I later can run queries and make statistics

When running this xml parse process I want to insert new data into both tables. So I'm creating new players and new teams in the database.

My question is, do I have to Insert the players into the database first and after that get their ids and insert them as a string array in the field members? 
Or how do I insert the team so that the relation with the users will work, is []string even the correct data type for the members field in the teams table? (Or does it has to be []Player?)

But if you say that using custom models is a bad practice for this case I will switch over go using records :)

**Comment by ganigeorgiev** [comment, marked as answer]
> do I have to Insert the players into the database first and after that get their ids and insert them as a string array in the field members?

Yes, that's usually the expected flow (_the integrity constraints are currently handled on application level because the relation is "dynamic", but this may change in the future_). Eventually you can wrap it inside a transaction with `app.Dao().RunInTransaction(...)`.

> how do I insert the team so that the relation with the users will work, is []string even the correct data type for the members field in the teams table? (Or does it has to be []Player?)

`[]string` will work out of the box only with `models.Record` because it normalizes (and cast to the correct type) any value set with `Set/Load` (internally it calls `schema/SchemaField.PrepareValue()`).
If you want to be able to save the multi-`relation` field values with a custom model struct, you'll need a datatype that implements `driver.Valuer` and that can be serialized into a json array. As mentioned in the above, an alternative could be to use the available `types.JsonArray` struct (it aliases `[]any` and it supports db read/write).


For the described use case, I personally will go with `models.Record` because it would be easier and doesn't require extra mapping. Here is an imaginary `processXml` function:

```go
func processXml(dao *daos.Dao, data map[string]any) error {
	playersCollection, err := dao.FindCollectionByNameOrId("players")
	if err != nil {
		return err
	}

	teamsCollection, err := dao.FindCollectionByNameOrId("teams")
	if err != nil {
		return err
	}

	return dao.RunInTransaction(func(txDao *daos.Dao) error {
		// iterate through the imaginary data...
		for k, v := range data {
			// create the members based on the iterated data
			members := []string{}
			for ... {
				player := models.NewRecord(playersCollection)
				player.Set("someField", someExtractedValue)
				if err := txDao.SaveRecord(player); err != nil {
					return err
				}
				members = append(members, player.Id)
			}

			// create new team
			// if you want to update, you can use any of the fetch helpers - https://pocketbase.io/docs/record-methods/#fetch-records
			team := models.NewRecord(teamsCollection)
			team.Set("members", members)
			team.Set("someOtherField", 123)
			if err := txDao.SaveRecord(team); err != nil {
				return err
			}
		}

		return nil
	})
}
```

**Reply by Jangerlp** [reply]
Thanks so much, this is exactly what I was looking for 🙂
